[Gradient Compression] PowerSGD comm hook

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48060 [Gradient Compression] PowerSGD comm hook**

Implement a PowerSGD variant that applies to a batched flattened tensor with zero paddings.

This version does not require handling 1D tensors and multi-dimenionsal tensors in the input separately, and hence it does not need to create two asyncrhonous future chains.

Potential optimization:
Consider FP16 compression throughout PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24843692](https://our.internmc.facebook.com/intern/diff/D24843692/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24843692/)!

Commit:
[Gradient Compression] PowerSGD comm hook.

Implement a PowerSGD variant that applies to a batched flattened tensor with zero paddings.

This version does not require handling 1D tensors and multi-dimenionsal tensors in the input separately, and hence it does not need to create two asyncrhonous future chains.

Potential optimization:
Consider FP16 compression throughout PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24843692](https://our.internmc.facebook.com/intern/diff/D24843692/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24843692/)!

[ghstack-poisoned]

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,109 @@
+    Initializes a 1D tensor from a standard normal distribution.
+    The see makes sure that the randomized vector in all the DDP replicas are the same at every step.
+    """
+    torch.manual_seed(np.random.RandomState(random_seed).randint(1_000_000_000))
```
If the np random number generator is only used once, I don't see the benefit of having it. You could just pass `random_seed` here directly. Maybe this shouldn't really be a separate function. You could just write something like:
```python
with torch.random.fork_rng():
    torch.manual_seed(0)  # if you do this multiple times, you should not repeatedly use the same seed
    vector.data[:] = torch.randn_like(vector)
```
in place of line 83 below.

SciPioneer replied:
Thanks!

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,109 @@
+    process_group: object, bucket: dist._GradBucket, rank: int = 1
+) -> torch.futures.Future:
+    """
+    This DDP communication hook implements a simplified PowerSGD graidient compression
```
typo: gradient compression

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,109 @@
+    algorithm described in https://arxiv.org/abs/1905.13727.
+    Once gradient tensors are aggregated across all workers, this hook applies
+    compression as follows:
+    1) Views the input flattened 1D gradient tensor as a square-shaped tensor M with 0 paddings;
```
I'm curious if this works. This would yield even stronger compression than regular rank-1/2 PowerSGD, where there are more smaller matrices that are not necessarily square.

SciPioneer replied:
The input tensor in my hook is actually a flattened 1D tensor that batches multiple per-variable tensors, corresponding to multiple neighboring layers in the network.

Here the input tensor does not need to have a square shape, and 0 paddings will be added to make a square.

The problem I found is that if I directly apply the original alg, if there are too many small tensors, the computation costs of orthogonizing and matrix multiplications over small matrices can be expensive. Here is more like a batching.

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,109 @@
+    p, p_memory = create_low_rank_tensor()
+    q, q_memory = create_low_rank_tensor()
+    # Initialize each Q from a standard normal distribution.
+    _set_random(q)
```
For each iteration, this should use a different 'seed'. 
I think it would be better to also orthogonalize Q here.

tvogels replied:
It seems that you implemented the algorithm without 'warm start'. Is that to simplify the initial tests?

SciPioneer replied:
> For each iteration, this should use a different 'seed'.
Now I used the length of input tensor as the seed. Added a TODO here to read it from an external state provided by the constructor of the hook.

> I think it would be better to also orthogonalize Q here.
Added this step. Curious: it seems that you commented out it in https://github.com/epfml/powersgd/blob/master/gradient_reducers.py#L677.



> It seems that you implemented the algorithm without 'warm start'. Is that to simplify the initial tests?
Correct. This is just the first version. I plan to add more optimizations iteratively.

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,109 @@
+    compression as follows:
+    1) Views the input flattened 1D gradient tensor as a square-shaped tensor M with 0 paddings;
+    2) Decomposes M into two low-rank tensors P and Q,
+    such that M = PQ^T, where Q is initialized from a standard normal distribution;
```
+ Q is orthogonalized

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,109 @@
+    4) Compute Q, which is approximately equal to M^TP;
+    5) Allreduces Q;
+    6) Computes M, which is approximately equal to PQ^T.
+    7) Truncates the input tensor to the original length.
```
This procedure (as described in the paper) does two matmul+allreduce steps per 'iteration': one left multiplication and one right multiplication. If you wanted, you could take one such step at a time, and alternate between them (for warm start)

SciPioneer replied:
Thanks for the advice! Are you referring to HalfRankKReducer implementation in your repo? Added a TODO here to try it in a different PR.

tvogels reviewed (commented):
@SciPioneer asked me to have a look at this. Looks good! Here are a few comments. Is the mechanism for error feedback somewhere else?

SciPioneer commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,109 @@
+        # If no epsilon is added here, division by zero may be caused by vanishing gradients.
+        # This epsilon is not needed if the input matrix covers the gradients of at least one entire layer in the neural network.
+        if epsilon > 0:
+            col += epsilon
```
Good catch! I somehow made a mistake when finally polishing the code.

I noticed that adding epsilon inline is faster than creating an intermediate denominator tensor.

Commit:
Update on "[Gradient Compression] PowerSGD comm hook."


Implement a PowerSGD variant that applies to a batched flattened tensor with zero paddings.

This version does not require handling 1D tensors and multi-dimenionsal tensors in the input separately, and hence it does not need to create two asyncrhonous future chains.

Potential optimization:
Consider FP16 compression throughout PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24843692](https://our.internmc.facebook.com/intern/diff/D24843692/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24843692/)!

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] PowerSGD comm hook."


Implement a PowerSGD variant that applies to a batched flattened tensor with zero paddings.

This version does not require handling 1D tensors and multi-dimenionsal tensors in the input separately, and hence it does not need to create two asyncrhonous future chains.

Potential optimization:
Consider FP16 compression throughout PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24843692](https://our.internmc.facebook.com/intern/diff/D24843692/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24843692/)!

[ghstack-poisoned]

SciPioneer commented:
> @SciPioneer asked me to have a look at this. Looks good! Here are a few comments. Is the mechanism for error feedback somewhere else?

Thanks for the comments. Haven't added error feedback yet.

martinjaggi commented:
Exiting to see this here!
About error feedback: Our experiments found it to be essential for good performance (Table 1 in the paper)

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,111 @@
+import torch.distributed as dist
+
+
+def _orthogonalize(matrix, epsilon=1e-8):
```
Any reason we're not using a built-in orthogonalization algorithm such as one provided by numpy that is battle-tested?

SciPioneer replied:
Let's use the paper author's implementation first, since I am not sure if there is any subtle difference between the two versions.

tvogels replied:
You could use `torch.qr()` and keep the `Q` part. That should do the same thing (apart from the epsilon parameter, maybe a good unit test) 
When I first wrote this, the hand-crafted procedure was faster. Maybe the QR algorithm is not optimized for long and narrow matrices with very few columns

SciPioneer replied:
> You could use `torch.qr()` and keep the `Q` part. That should do the same thing (apart from the epsilon parameter, maybe a good unit test)

In this implementation, epsilon is 0 when orthogonalizing both P and Q. Why do you suggest keeping the Q part?

> When I first wrote this, the hand-crafted procedure was faster. Maybe the QR algorithm is not optimized for long and narrow matrices with very few columns

Interesting! Since we expect the rank here is a very small number, typically 1 or 2, then probably it's better to go with your hand-crafted version for now.

tvogels replied:
> Why do you suggest keeping the Q part?
Ah, with Q I meant the Q from Q-R decomposition, not the Q from the P and Q in PowerSGD.

SciPioneer replied:
> > Why do you suggest keeping the Q part?
> 
> Ah, with Q I meant the Q from Q-R decomposition, not the Q from the P and Q in PowerSGD.

Thanks! I think you meant `matrix[:], _ = torch.qr(matrix)`

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,111 @@
+                # TODO(wayi@): Should read the random seed from the state of this hook provided by the constructor.
+                torch.manual_seed(total_length)
+                memory[:] = torch.randn_like(memory)
+        return memory.view(square_side_length, rank), memory
```
I know this is a low-rank matrix approximation algorithm, but can we somehow distinguish rank for low-rank and rank for process group identifier? Could get very confusing easily if we don't do this.

rohan-varma replied:
Should `rank` be an input into this function rather than implicitly captured from the context? Also, should we have unittests for this function?

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,111 @@
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def powerSGD_hook(
```
Would a C++ implementation of this hook be more efficient?

rohan-varma replied:
Should we have thorough unit testing for this hook?

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,111 @@
+def powerSGD_hook(
+    process_group: object, bucket: dist._GradBucket, rank: int = 1
+) -> torch.futures.Future:
+    """
```
Nit: the docstring should probably specify what each arg is and the default that is used.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,111 @@
+    2) Decomposes M into two low-rank tensors P and Q,
+    such that M = PQ^T, where Q is initialized from a standard normal distribution and orthogonalized;
+    2) Allreduces P;
+    3) Orthogonizes P;
```
nit: it's "orthogonalizes" right?

SciPioneer replied:
It's orthogonalized. See https://github.com/pytorch/pytorch/pull/48060#discussion_r525092833

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,111 @@
+def _orthogonalize(matrix, epsilon=1e-8):
+    """Applies Gram-Schmidt procedure to orthogonalize a given 2D tensor."""
+    num_cols = matrix.shape[1]
+    for i in range(num_cols):
```
This appears like a procedure that's only dependent on the ith col of the matrix, shouldn't we be able to use some sort of parallelism here (`at::parallel_for` comes to mind, but I'm not an expert on PT threading and parallelism) to speed this procedure up if it were implemented in C++ ?

SciPioneer replied:
Actually there is only 1 column by default, and at most 2 columns at most for good compression performance.

tvogels replied:
The for-loop is unfortunately inherently sequential here. Each iteration changes the data that will be used by the next one.

tvogels replied:
If the compression performance (speed) hinges on this orthogonalization step, it may be a good idea to try a C++ implementation as @rohan-varma suggested above. Not sure if rank 2 would be high enough *in general* to train a model to full accuracy in the same number of steps as with full-precision communication.

SciPioneer replied:
If the rank is high, the column-wise scanning can hit a performance issue even implemented in C++, as it won't be cache friendly. Then I think we can further investigate your follow-up work that runs rank-1 compression multiple times.

For now I want to see the max speedup of PowerSGD by using the most aggressive compression setup.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,111 @@
+                memory[:] = torch.randn_like(memory)
+        return memory.view(square_side_length, rank), memory
+
+    p, p_memory = create_low_rank_tensor(fill_random_values=False)
```
May be lack of understanding on my part, but why do we need the tensor and the allocated contiguous memory?

tvogels replied:
This is probably not necessary in the code here. This implementation is variation on the original. In the original version, multiple tensors were stored in a contiguous chunk of memory. The individual tensors are computed separately from different parameters of the network, but the communication happened on the one contiguous buffer.

Commit:
Update on "[Gradient Compression] PowerSGD comm hook"


Implement a PowerSGD variant that applies to a batched flattened tensor with zero paddings.

This version does not require handling 1D tensors and multi-dimenionsal tensors in the input separately, and hence it does not need to create two asyncrhonous future chains.

Potential optimization:
Consider FP16 compression throughout PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24843692](https://our.internmc.facebook.com/intern/diff/D24843692/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24843692/)!

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] PowerSGD comm hook"


Implement a PowerSGD variant that applies to a batched flattened tensor with zero paddings.

This version does not require handling 1D tensors and multi-dimenionsal tensors in the input separately, and hence it does not need to create two asyncrhonous future chains.

Potential optimization:
Consider FP16 compression throughout PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24843692](https://our.internmc.facebook.com/intern/diff/D24843692/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24843692/)!

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] PowerSGD comm hook"


Implement a PowerSGD variant that applies to a batched flattened tensor with zero paddings.

This version does not require handling 1D tensors and multi-dimenionsal tensors in the input separately, and hence it does not need to create two asyncrhonous future chains.

Potential optimization:
Consider FP16 compression throughout PowerSGD.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D24843692](https://our.internmc.facebook.com/intern/diff/D24843692/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24843692/)!

[ghstack-poisoned]

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -0,0 +1,111 @@
+                # Currently use the length of input tensor as the seed, which should be mostly different.
+                # TODO(wayi@): Should read the random seed from the state of this hook provided by the constructor.
+                torch.manual_seed(total_length)
+                memory[:] = torch.randn_like(memory)
```
If you are already optimizing for compression speed, this random number generation step seems to quite expensive. For that reason, the `warm start` alternative, which uses state instead of fresh random numbers is faster. You could also consider pre-generate a tensor of random numbers and select chunks from there. 

Note that you should use fresh random numbers here for each call of the hook (Is that what the TODO says?). `total_length` seems to be fixed.

SciPioneer replied:
> If you are already optimizing for compression speed, this random number generation step seems to quite expensive. For that reason, the `warm start` alternative, which uses state instead of fresh random numbers is faster. You could also consider pre-generate a tensor of random numbers and select chunks from there.

This version is a bit outdated. Actually I added the `torch.device` to `fork_rng()`, so rng only runs on a single GPU core at a time. I did run another experiment to evaluate the speed, and it seems that such random generator is not too slow.

> Note that you should use fresh random numbers here for each call of the hook (Is that what the TODO says?). `total_length` seems to be fixed.

Thanks for the reminder! `total_length` is the length of input tensor, and for now it is different for each call of the hook (at the same iteration), because the models that I am experimenting on has many dense layers of very different sizes. This assumption may not hold true for CNN.

TODO says that I should pass in a random generator, and use a generated random value as the seed every time. This is the first version of PowerSGD for experimentations, and I don't plan to introduce any extra state for the hook for now.

This means that this version of the hook is stateless, and it does not support warm-start or error feedback. I first want to make sure PowerSGD is promising performance-wise in our setup.

tvogels replied:
Make sense. Thanks for the explanation. When interpreting the outcomes of this test, you should keep in mind that using *different* 'projections' of the gradients at different SGD steps is quite crucial. If you use the same random projection every time, there will be differences between the worker's gradients that are never synchronized.
